### PR TITLE
MAINT: Tidy exception chaining in  _datasource.py

### PR DESCRIPTION
**Pull request by sarthakvk**
this solution is related to the following issue #15986
@eric-wieser this is my first contribution, can you review this pr?
Thank you.

### CODE REVIEW

File: numpy/lib/_datasource.py [hunk 1]
```diff
@@ -337,8 +337,8 @@ def _cache(self, path):
                 with closing(urlopen(path)) as openedurl:
                     with _open(upath, 'wb') as f:
                         shutil.copyfileobj(openedurl, f)
-            except URLError:
-                raise URLError("URL not found: %s" % path)
+            except URLError as e:
```
Comment (eric-wieser) [inline numpy/lib/_datasource.py:340]: Can you show what the resulting error message is with this change?

Comment (sarthakchaudhary13) [reply numpy/lib/_datasource.py:340]: - *Before invalid URL exeption ( I run this on my personal conda environment ) was*

```
Traceback (most recent call last):
  File "/home/sarthak/anaconda3/lib/python3.7/site-packages/numpy/lib/_datasource.py", line 418, in _cache
    with closing(urlopen(path)) as openedurl:
  File "/home/sarthak/anaconda3/lib/python3.7/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/home/sarthak/anaconda3/lib/python3.7/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/home/sarthak/anaconda3/lib/python3.7/urllib/request.py", line 641, in http_response
    'http', request, response, code, msg, hdrs)
  File "/home/sarthak/anaconda3/lib/python3.7/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/home/sarthak/anaconda3/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/home/sarthak/anaconda3/lib/python3.7/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sarthak/anaconda3/lib/python3.7/site-packages/numpy/lib/_datasource.py", line 422, in _cache
    raise URLError("URL not found: %s" % path)
urllib.error.URLError: <urlopen error URL not found: https://raw.githubusercontent.com/numpy/numpy/master/toxsadas.ini>
```
- *After the change*
```
Traceback (most recent call last):
  File "/home/sarthak/Desktop/projects/numpy/numpy/lib/_datasource.py", line 337, in _cache
    with closing(urlopen(path)) as openedurl:
  File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sarthak/Desktop/projects/numpy/numpy/lib/_datasource.py", line 341, in _cache
    raise URLError("URL not found: %s" % path) from e
urllib.error.URLError: <urlopen error URL not found: https://raw.githubusercontent.com/numpy/numpy/master/toxsadas.ini>
```

**Review by eric-wieser — APPROVED** [review]
Definitely an improvement, but I wonder if we just want to use the inner exception by itself, and remove the try/except. Something for a follow-up PR, perhaps.

**Comment by sarthakchaudhary13** [comment]
> Definitely an improvement, but I wonder if we just want to use the inner exception by itself, and remove the try/except. Something for a follow-up PR, perhaps.

@eric-wieser  Correct me if i am wrong :confused:  , the error should look something like
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sarthak/Desktop/projects/numpy/numpy/lib/_datasource.py", line 336, in _cache
    with closing(urlopen(path)) as openedurl:
  File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found
```

**Review by eric-wieser — APPROVED** [review]
Great, thanks - looking back at the history, it looks like the second error message was around forever (0ce4792257d812a05ce222d0e5b5eb1aa54d19fd) and probably wasn't very deliberate.

File: numpy/lib/_datasource.py [hunk 2]
```diff
@@ -333,12 +333,9 @@ def _cache(self, path):
 
         # TODO: Doesn't handle compressed files!
         if self._isurl(path):
-            try:
-                with closing(urlopen(path)) as openedurl:
-                    with _open(upath, 'wb') as f:
-                        shutil.copyfileobj(openedurl, f)
-            except URLError:
-                raise URLError("URL not found: %s" % path)
+            with closing(urlopen(path)) as openedurl:
```
Comment (eric-wieser) [inline numpy/lib/_datasource.py:336]: I bet `closing` isn't needed any more either...

**Comment by sarthakchaudhary13** [comment]
> 
> ```
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/sarthak/Desktop/projects/numpy/numpy/lib/_datasource.py", line 336, in _cache
>     with closing(urlopen(path)) as openedurl:
>   File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
>     return opener.open(url, data, timeout)
>   File "/usr/lib/python3.8/urllib/request.py", line 531, in open
>     response = meth(req, response)
>   File "/usr/lib/python3.8/urllib/request.py", line 640, in http_response
>     response = self.parent.error(
>   File "/usr/lib/python3.8/urllib/request.py", line 569, in error
>     return self._call_chain(*args)
>   File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
>     result = func(*args)
>   File "/usr/lib/python3.8/urllib/request.py", line 649, in http_error_default
>     raise HTTPError(req.full_url, code, msg, hdrs, fp)
> urllib.error.HTTPError: HTTP Error 404: Not Found
> ```

@eric-wieser this is the error after the change.

File: numpy/lib/_datasource.py (hunk 2 above)
Comment (sarthakchaudhary13) [reply numpy/lib/_datasource.py:336]: Yes, making another commit.
Thank you

File: numpy/lib/_datasource.py [hunk 3]
```diff
@@ -333,12 +332,9 @@ def _cache(self, path):
 
         # TODO: Doesn't handle compressed files!
         if self._isurl(path):
-            try:
-                with closing(urlopen(path)) as openedurl:
-                    with _open(upath, 'wb') as f:
-                        shutil.copyfileobj(openedurl, f)
-            except URLError:
-                raise URLError("URL not found: %s" % path)
+            openedurl = urlopen(path)
```
Comment (eric-wieser) [inline numpy/lib/_datasource.py:335]: The `with` statement is still needed, even if `closing` is not.

Comment (sarthakchaudhary13) [reply numpy/lib/_datasource.py]: I'm sorry :sweat: , is it fine now?

**Review by eric-wieser — APPROVED** [review]

**Comment by rossbar** [comment]
Thanks @sarthakchaudhary13 and @eric-wieser for the review!

**Comment by sarthakchaudhary13** [comment]
@eric-wieser and @rossbar thanks for helping me making my first contribution in numpy I am hoping to do more :smile:

**Pull request merged**